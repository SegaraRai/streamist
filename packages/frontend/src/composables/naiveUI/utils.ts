import type { MenuOption } from 'naive-ui';

export function cleanupDividers(
  menuOptions: readonly MenuOption[]
): MenuOption[] {
  const filtered = menuOptions.filter(
    (option, index, arr) =>
      option.type !== 'divider' ||
      (index !== 0 && arr[index - 1].type !== 'divider')
  );
  if (filtered[filtered.length - 1]?.type === 'divider') {
    filtered.pop();
  }
  return filtered;
}
