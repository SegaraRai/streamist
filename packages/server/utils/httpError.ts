export interface HTTPErrorData {
  readonly message: string;
}

export class HTTPError extends Error {
  readonly status: number;
  readonly data: HTTPErrorData;

  constructor(status: number, message: string) {
    super(`${status} ${message}`);

    this.status = status;
    this.data = {
      message,
    };
  }
}
