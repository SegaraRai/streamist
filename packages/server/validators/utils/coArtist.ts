import {
  ValidationArguments,
  ValidationOptions,
  ValidatorConstraint,
  ValidatorConstraintInterface,
  registerDecorator,
} from 'class-validator';
import { isValidCoArtistRole } from '$shared/coArtist';

// class-validatorはv0.13.2時点でカスタムデコレーターに対するeachオプションに対応していないため、
// それらについては@IsIdデコレーターではなく@ValidateデコレーターとIsIdConstraintを組み合わせて検証する
// https://github.com/typestack/class-validator/issues/237
@ValidatorConstraint({
  name: 'isCoArtistRoleConstraint',
  async: false,
})
export class IsCoArtistRoleConstraint implements ValidatorConstraintInterface {
  validate(value: unknown): boolean {
    return isValidCoArtistRole(value);
  }

  defaultMessage(args: ValidationArguments): string {
    return `${args.property} must be a valid coArtist role`;
  }
}

export function IsCoArtistRole(
  validationOptions?: ValidationOptions
): PropertyDecorator {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return function (object: Record<string, any>, propertyName: string): void {
    registerDecorator({
      name: 'isCoArtistRole',
      target: object.constructor,
      propertyName,
      constraints: [],
      options: validationOptions,
      validator: IsCoArtistRoleConstraint,
    });
  } as PropertyDecorator;
}
