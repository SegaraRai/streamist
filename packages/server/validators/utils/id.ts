import {
  ValidationArguments,
  ValidationOptions,
  ValidatorConstraint,
  ValidatorConstraintInterface,
  registerDecorator,
} from 'class-validator';
import { isId } from '$shared/id';

// class-validatorはv0.13.2時点でカスタムデコレーターに対するeachオプションに対応していないため、
// それらについては@IsIdデコレーターではなく@ValidateデコレーターとIsIdConstraintを組み合わせて検証する
// https://github.com/typestack/class-validator/issues/237
@ValidatorConstraint({
  name: 'isIdConstraint',
  async: false,
})
export class IsIdConstraint implements ValidatorConstraintInterface {
  validate(value: unknown): boolean {
    return isId(value);
  }

  defaultMessage(args: ValidationArguments): string {
    return `${args.property} must be a valid id`;
  }
}

export function IsId(validationOptions?: ValidationOptions): PropertyDecorator {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return function (object: Record<string, any>, propertyName: string): void {
    registerDecorator({
      name: 'isId',
      target: object.constructor,
      propertyName,
      constraints: [],
      options: validationOptions,
      validator: IsIdConstraint,
    });
  } as PropertyDecorator;
}
