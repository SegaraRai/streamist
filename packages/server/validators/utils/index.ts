export * from './id';
export * from './region';
export * from './optional';
