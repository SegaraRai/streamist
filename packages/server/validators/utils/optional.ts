import { ValidateIf, ValidationOptions } from 'class-validator';

export function IsNullable(
  this: unknown,
  options?: ValidationOptions
): PropertyDecorator {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return ValidateIf.call(
    this,
    (_object, value): boolean => value !== null,
    options
  );
}

export function IsUndefinable(
  this: unknown,
  options?: ValidationOptions
): PropertyDecorator {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return ValidateIf.call(
    this,
    (_object, value): boolean => value !== undefined,
    options
  );
}
