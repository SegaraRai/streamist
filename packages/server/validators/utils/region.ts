import {
  ValidationArguments,
  ValidationOptions,
  ValidatorConstraint,
  ValidatorConstraintInterface,
  registerDecorator,
} from 'class-validator';
import { isRegion } from '$shared/regions';

// class-validatorはv0.13.2時点でカスタムデコレーターに対するeachオプションに対応していないため、
// それらについては@IsRegionデコレーターではなく@ValidateデコレーターとIsRegionConstraintを組み合わせて検証する
// https://github.com/typestack/class-validator/issues/237
@ValidatorConstraint({
  name: 'isRegionConstraint',
  async: false,
})
export class IsRegionConstraint implements ValidatorConstraintInterface {
  validate(value: unknown): boolean {
    return isRegion(value);
  }

  defaultMessage(args: ValidationArguments): string {
    return `${args.property} must be a valid region`;
  }
}

export function IsRegion(
  validationOptions?: ValidationOptions
): PropertyDecorator {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return function (object: Record<string, any>, propertyName: string): void {
    registerDecorator({
      name: 'isRegion',
      target: object.constructor,
      propertyName,
      constraints: [],
      options: validationOptions,
      validator: IsRegionConstraint,
    });
  } as PropertyDecorator;
}
