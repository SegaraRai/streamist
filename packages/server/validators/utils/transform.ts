import {
  normalizeTextForMultipleLines,
  normalizeTextForSingleLine,
} from '$shared/normalize';

export function tStringNormalizeSingleLine(value: unknown): unknown {
  return typeof value === 'string'
    ? normalizeTextForSingleLine(value) || ''
    : value;
}

export function tStringNormalizeMultipleLines(value: unknown): unknown {
  return typeof value === 'string'
    ? normalizeTextForMultipleLines(value) || ''
    : value;
}
