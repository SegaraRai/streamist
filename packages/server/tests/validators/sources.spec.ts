import { describe, expect, it } from 'vitest';
import {
  VSourceCreateBodyWrapper,
  VSourceFinishUploadBody,
} from '$/validators';
import { validate } from './utils';

describe('VSourceCreateBodyWrapper', () => {
  it('should accept correct payloads', async () => {
    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: 'audio',
          audioFile: {
            type: 'audio',
            filename: 'test.mp3',
            fileSize: 1000,
          },
          cueSheetFile: null,
          region: 'ap-northeast-1',
        },
      })
    ).resolves.toBeUndefined();

    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: 'audio',
          audioFile: {
            type: 'audio',
            filename: 'test.wav',
            fileSize: 1000,
          },
          cueSheetFile: {
            type: 'cueSheet',
            filename: 'test.cue',
            fileSize: 200,
          },
          region: 'ap-northeast-1',
        },
      })
    ).resolves.toBeUndefined();

    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: 'image',
          imageFile: {
            type: 'image',
            filename: 'test.jpg',
            fileSize: 1000,
          },
          attachToType: 'album',
          attachToId: 'test',
          attachPrepend: true,
          region: 'ap-northeast-1',
        },
      })
    ).resolves.toBeUndefined();
  });

  it('should reject malformed payloads', async () => {
    // invalid region
    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: 'audio',
          audioFile: {
            type: 'audio',
            filename: 'test.mp3',
            fileSize: 1000,
          },
          cueSheetFile: null,
          region: 'ap-northeast-x' as any,
        },
      })
    ).rejects.toMatchSnapshot();

    // invalid id
    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: 'image',
          imageFile: {
            type: 'image',
            filename: 'test.jpg',
            fileSize: 1000,
          },
          attachToType: 'album',
          attachToId: '/test',
          attachPrepend: true,
          region: 'ap-northeast-1',
        },
      })
    ).rejects.toMatchSnapshot();

    // invalid file type
    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: 'audio',
          audioFile: {
            type: 'image' as any,
            filename: 'test.mp3',
            fileSize: 1000,
          },
          cueSheetFile: null,
          region: 'ap-northeast-1',
        },
      })
    ).rejects.toMatchSnapshot();

    // inconsistent type
    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: 'image' as any,
          audioFile: {
            type: 'audio' as any,
            filename: 'test.mp3',
            fileSize: 1000,
          },
          cueSheetFile: null,
          region: 'ap-northeast-1',
        },
      })
    ).rejects.toMatchSnapshot();

    // invalid type
    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: 'foo' as any,
          audioFile: {
            type: 'audio',
            filename: 'test.mp3',
            fileSize: 1000,
          },
          cueSheetFile: null,
          region: 'ap-northeast-1',
        },
      })
    ).rejects.toMatchSnapshot();

    // invalid type of type
    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          type: ['audio'] as any,
          audioFile: {
            type: 'audio',
            filename: 'test.mp3',
            fileSize: 1000,
          },
          cueSheetFile: null,
          region: 'ap-northeast-1',
        },
      })
    ).rejects.toMatchSnapshot();

    // missing type
    await expect(
      validate(VSourceCreateBodyWrapper, {
        '!payload': {
          audioFile: {
            type: 'audio',
            filename: 'test.mp3',
            fileSize: 1000,
          },
          cueSheetFile: null,
          region: 'ap-northeast-1',
        } as any,
      })
    ).rejects.toMatchSnapshot();
  });
});

describe('VSourceFinishUploadBody', () => {
  it('should accept correct payloads', async () => {
    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploaded',
      })
    ).resolves.toBeUndefined();

    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploaded',
        parts: undefined,
      })
    ).resolves.toBeUndefined();

    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploaded',
        parts: ['etag1'],
      })
    ).resolves.toBeUndefined();

    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploaded',
        parts: ['etag1', 'etag2'],
      })
    ).resolves.toBeUndefined();
  });

  it('should reject malformed payloads', async () => {
    // invalid state
    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploadable' as any,
      })
    ).rejects.toMatchSnapshot();

    // null parts
    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploaded',
        parts: null as any,
      })
    ).rejects.toMatchSnapshot();

    // empty parts
    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploaded',
        parts: [],
      })
    ).rejects.toMatchSnapshot();

    // invalid type of parts
    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploaded',
        parts: [123] as any,
      })
    ).rejects.toMatchSnapshot();

    // object parts
    await expect(
      validate(VSourceFinishUploadBody, {
        state: 'uploaded',
        parts: { 0: 'etag1' } as any,
      })
    ).rejects.toMatchSnapshot();
  });
});
