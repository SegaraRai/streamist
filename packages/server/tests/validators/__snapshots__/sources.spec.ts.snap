// Vitest Snapshot v1

exports[`VSourceCreateBodyWrapper > should reject malformed payloads 1`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isRegionConstraint": "region must be a valid region",
        },
        "property": "region",
        "value": "ap-northeast-x",
      },
    ],
    "property": "!payload",
    "value": VSourceCreateBodyAudio {
      "audioFile": VSourceCreateBodyAudioFile {
        "fileSize": 1000,
        "filename": "test.mp3",
        "type": "audio",
      },
      "cueSheetFile": null,
      "region": "ap-northeast-x",
      "type": "audio",
    },
  },
]
`;

exports[`VSourceCreateBodyWrapper > should reject malformed payloads 2`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isIdConstraint": "attachToId must be a valid id",
        },
        "property": "attachToId",
        "value": "/test",
      },
    ],
    "property": "!payload",
    "value": VSourceCreateBodyImage {
      "attachPrepend": true,
      "attachToId": "/test",
      "attachToType": "album",
      "imageFile": VSourceCreateBodyImageFile {
        "fileSize": 1000,
        "filename": "test.jpg",
        "type": "image",
      },
      "region": "ap-northeast-1",
      "type": "image",
    },
  },
]
`;

exports[`VSourceCreateBodyWrapper > should reject malformed payloads 3`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [
          ValidationError {
            "children": [],
            "constraints": {
              "equals": "type must be equal to audio",
            },
            "property": "type",
            "value": "image",
          },
        ],
        "property": "audioFile",
        "value": VSourceCreateBodyAudioFile {
          "fileSize": 1000,
          "filename": "test.mp3",
          "type": "image",
        },
      },
    ],
    "property": "!payload",
    "value": VSourceCreateBodyAudio {
      "audioFile": VSourceCreateBodyAudioFile {
        "fileSize": 1000,
        "filename": "test.mp3",
        "type": "image",
      },
      "cueSheetFile": null,
      "region": "ap-northeast-1",
      "type": "audio",
    },
  },
]
`;

exports[`VSourceCreateBodyWrapper > should reject malformed payloads 4`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isIn": "attachToType must be one of the following values: album, artist, playlist",
        },
        "property": "attachToType",
        "value": undefined,
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isIdConstraint": "attachToId must be a valid id",
        },
        "property": "attachToId",
        "value": undefined,
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isBoolean": "attachPrepend must be a boolean value",
        },
        "property": "attachPrepend",
        "value": undefined,
      },
    ],
    "property": "!payload",
    "value": VSourceCreateBodyImage {
      "attachPrepend": undefined,
      "attachToId": undefined,
      "attachToType": undefined,
      "audioFile": {
        "fileSize": 1000,
        "filename": "test.mp3",
        "type": "audio",
      },
      "cueSheetFile": null,
      "imageFile": undefined,
      "region": "ap-northeast-1",
      "type": "image",
    },
  },
]
`;

exports[`VSourceCreateBodyWrapper > should reject malformed payloads 5`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "constraints": {
          "nestedValidation": "nested property !payload must be either object or array",
        },
        "property": "!payload",
        "target": [Function],
        "value": true,
      },
    ],
    "constraints": {
      "isObject": "type must be either \\"audio\\" or \\"image\\"",
    },
    "property": "!payload",
    "value": true,
  },
]
`;

exports[`VSourceCreateBodyWrapper > should reject malformed payloads 6`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "constraints": {
          "nestedValidation": "nested property !payload must be either object or array",
        },
        "property": "!payload",
        "target": [Function],
        "value": true,
      },
    ],
    "constraints": {
      "isObject": "type must be either \\"audio\\" or \\"image\\"",
    },
    "property": "!payload",
    "value": true,
  },
]
`;

exports[`VSourceCreateBodyWrapper > should reject malformed payloads 7`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "constraints": {
          "nestedValidation": "nested property !payload must be either object or array",
        },
        "property": "!payload",
        "target": [Function],
        "value": true,
      },
    ],
    "constraints": {
      "isObject": "type must be either \\"audio\\" or \\"image\\"",
    },
    "property": "!payload",
    "value": true,
  },
]
`;

exports[`VSourceFinishUploadBody > should reject malformed payloads 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isIn": "state must be one of the following values: aborted, uploaded",
    },
    "property": "state",
    "value": "uploadable",
  },
]
`;

exports[`VSourceFinishUploadBody > should reject malformed payloads 2`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "arrayNotEmpty": "parts should not be empty",
      "isArray": "parts must be an array",
      "isString": "each value in parts must be a string",
    },
    "property": "parts",
    "value": null,
  },
]
`;

exports[`VSourceFinishUploadBody > should reject malformed payloads 3`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "arrayNotEmpty": "parts should not be empty",
    },
    "property": "parts",
    "value": [],
  },
]
`;

exports[`VSourceFinishUploadBody > should reject malformed payloads 4`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "each value in parts must be a string",
    },
    "property": "parts",
    "value": [
      123,
    ],
  },
]
`;

exports[`VSourceFinishUploadBody > should reject malformed payloads 5`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "arrayNotEmpty": "parts should not be empty",
      "isArray": "parts must be an array",
      "isString": "each value in parts must be a string",
    },
    "property": "parts",
    "value": {
      "0": "etag1",
    },
  },
]
`;
